CriteoPublisherSdk/Sources/Logging/CR_RemoteLogRecordSerializer.m

```diff
@@ -0,0 +1,67 @@
+//
+//  CR_RemoteLogRecordSerializer.m
+//  CriteoPublisherSdk
+//
+//  Copyright © 2018-2021 Criteo. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "CR_RemoteLogRecordSerializer.h"
+
+@implementation CR_RemoteLogRecordSerializer
+
+- (NSArray<NSDictionary<NSString *, NSObject *> *> *)bodyForRemoteLogRecords:
+    (NSArray<CR_RemoteLogRecord *> *)records {
+  NSMutableArray *formattedArray = [NSMutableArray arrayWithCapacity:records.count];
+  for (CR_RemoteLogRecord *record in records) {
+    NSDictionary<NSString *, NSObject *> *dictionary = [self dictionaryForRemoteLogRecord:record];
+    [formattedArray addObject:dictionary];
+  }
+  return formattedArray;
+}
+
+- (NSDictionary<NSString *, NSObject *> *)dictionaryForRemoteLogRecord:
+    (CR_RemoteLogRecord *)record {
+#pragma clang diagnostic push
+#pragma ide diagnostic ignored "OCLegacyObjCLiteralInspection"
+  // The exception type is nullable, so it is not possible to use @{} notation.
+  NSDictionary *context = [NSDictionary
+      dictionaryWithObjectsAndKeys:record.version, @"version", record.bundleId, @"bundleId",
+                                   record.deviceId, @"deviceId", record.sessionId, @"sessionId",
+                                   record.profileId, @"profileId", record.tag, @"tag",
+                                   record.exceptionType, @"exception", nil];
+#pragma clang diagnostic pop
+
+  return @{
+    @"context" : context,
+    @"errors" : @[
+      @{@"errorType" : [self formatSeverity:record.severity], @"messages" : @[ record.message ]}
+    ]
+  };
+}
+
+- (NSString *)formatSeverity:(CR_LogSeverity)severity {
+  switch (severity) {
+    case CR_LogSeverityError:
+      return @"Error";
+    case CR_LogSeverityWarning:
+      return @"Warning";
+    case CR_LogSeverityInfo:
+      return @"Info";
+    default:
```

Comment:
NIT: I would rather avoid having a default there to enforce a developer to amend this if levels were to be added. Unless there is a consequence on backend.

Reply:
If level is not known by backend (Debug, Info, Warning or Error), then the backend will just throw out the whole request. Not the best behavior, but at least, it will not make the server throw.

Let me remove the default then.